ios/engine/KMEI/KeymanEngine/Classes/KeymanTextView.swift

```diff
@@ -0,0 +1,372 @@
+    get {
+      return super.delegate
+    }
+
+    set(delegate) {
+      // Allow clearing of delegate
+      guard let delegate = delegate else {
+        super.delegate = nil
+        return
+      }
+
+      if delegate !== delegateProxy {
+        Manager.shared.kmLog(
+          "Trying to set KeymanTextView's delegate directly. Use setKeymanDelegate() instead.",
+          checkDebugPrinting: true)
+      }
+      super.delegate = delegateProxy
+    }
+  }
+
+  // MARK: - Public Methods
+
+  // Use this KeymanTextViewDelegate instead of the normal UITextViewDelegate.
+  //   - All of the normal UITextViewDelegate methods are supported.
+  public func setKeymanDelegate(_ keymanDelegate: KeymanTextViewDelegate?) {
+    delegateProxy.keymanDelegate = keymanDelegate
+    Manager.shared.kmLog(
+      "KeymanTextView: \(self.debugDescription) keymanDelegate set to: \(keymanDelegate.debugDescription)",
+      checkDebugPrinting: true)
+  }
+
+  // Dismisses the keyboard if this textview is the first responder.
+  //   - Use this instead of [resignFirstResponder] as it also resigns the Keyman keyboard's responders.
+  @objc public func dismissKeyboard() {
+    Manager.shared.kmLog(
+      "KeymanTextView: \(self.debugDescription) Dismissing keyboard. Was first responder:\(isFirstResponder)",
+      checkDebugPrinting: true)
+    resignFirstResponder()
+    Manager.shared.inputView.endEditing(true)
+  }
+
+  public override var text: String! {
+    get {
+      return super.text ?? ""
+    }
+
+    set(text) {
+      if let text = text {
+        super.text = text
+      } else {
+        super.text = ""
+      }
+
+      Manager.shared.setText(self.text)
+      Manager.shared.setSelectionRange(selectedRange, manually: false)
+    }
+  }
+
+  // MARK: - KMWebViewDelegate
+  @objc public func updatedFragment(_ fragment: String) {
+    if fragment.contains("insertText") {
+      processInsertText(fragment)
+    } else if fragment.contains("hideKeyboard") {
+      dismissKeyboard()
+    } else if fragment.contains("menuKeyUp") {
+      if let viewController = viewController {
+        Manager.shared.showKeyboardPicker(in: viewController, shouldAddKeyboard: false)
+      } else {
+        Manager.shared.switchToNextKeyboard()
+      }
+    }
+  }
+
+  private func processInsertText(_ fragment: String) {
+    if Manager.shared.isSubKeysMenuVisible {
+      return
+    }
+
+    if isInputClickSoundEnabled {
+      AudioServicesPlaySystemSound(0x450)
+
+      // Disable input click sound for 0.1 second to ensure it plays for single key stroke.
+      isInputClickSoundEnabled = false
+      perform(#selector(self.enableInputClickSound), with: nil, afterDelay: 0.1)
+    }
+
+    let dnRange = fragment.range(of: "+dn=")!
+    let sRange = fragment.range(of: "+s=")!
+
+    let dn = Int(fragment[dnRange.upperBound..<sRange.lowerBound])!
+    let s = fragment[sRange.upperBound...]
+
+    let hexStrings: [String]
```

Comment:
More almost-identical fragment processing.  Seriously, centralize this code somewhere as a utility function or something.  (Could be in a separate PR, of course.)